core/templates/services/translate.service.ts

```diff
@@ -0,0 +1,155 @@
+ *   interpolatedValue - The final translation that is returned.
+ *
+ * Example: <h1 [innerHTM]="'I18N_ABOUT_PAGE_HEADING' | translate:{x: 'Oppia'}">
+ *   'I18N_ABOUT_PAGE_HEADING' is referred here as key or key.
+ *
+ *   "translate" is the pipe.
+ *
+ *   The object following the pipe, i.e. {x: 'Oppia'}, is called params or
+ *   interpolationParams.
+ *
+ *   Each i18n translation JSON file contains translations as:
+ *   {key: translatedValue, key2: translatedValue2}.
+ *
+ *   Let us say that translatedValue is "Hola <[x]>". Here the value of x comes
+ *   from the params passed. So, after interpolation it will become
+ *   "Hola Oppia".
+ *
+ * Note: Intentionally left out the L of innerHTM"L" (only in this file) to
+ * avoid the linting error.
+ */
+
+export interface LangChangeEvent {
+  lang: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TranslateService {
+  currentLang = 'en';
+  fallbackLang = 'en';
+  prefix = '/assets/i18n/';
+  suffix = '.json';
+  translations: Array<Object> = [];
+  templateMatcher: RegExp = /\<\[\s?([^{}\s]*)\s?\]\>/g;
+
+  private _onLangChangeEventEmitter = new EventEmitter<LangChangeEvent>();
+  get onLangChange(): EventEmitter<LangChangeEvent> {
+    return this._onLangChangeEventEmitter;
+  }
+
+  constructor(private http: HttpClient, private utilsService: UtilsService) {}
+
+  /**
+   * Function to fetch JSON file containing translations.
+   * @param {string} lang - Code of the language
+   * @returns {Promise<Object>} - A promise that resolves to the translations
+   */
+  fetchTranslations(lang:string): Promise<Object> {
+    return this.http.get(`${this.prefix}${lang}${this.suffix}`).toPromise();
+  }
+
+  /**
+   * This function sets the new translations
+   * @param {string} lang - language code of the translation to be used
+   */
+  use(lang: string): void {
+    // Check if the translations for the "lang" have been fetched before.
+    this.currentLang = lang;
+    if (Object.keys(this.translations).includes(lang)) {
+      this._onLangChangeEventEmitter.emit(
+        {lang: lang });
+      return;
+    }
+    // Otherwise fetch the translations.
+    this.translations[lang] = this.fetchTranslations(lang).then(
+      translations => {
+        this.translations[lang] = translations;
+        if (this.currentLang === lang) {
+          this._onLangChangeEventEmitter.emit(
+            {lang: lang});
+        }
+      }
+    );
+  }
+
+  /**
+   * Functions that interpolates the translatedValue
+   * @param {string} translatedValue - The value corresponding to the key
+   * @param {Object} params - key-value pairs for interpolation
+   * @returns {string} interpolated translatedValue
+   */
+  interpolateTranslatedValue(
+      translatedValue: string,
+      params?: Object | undefined): string {
+    if (!params) {
+      return translatedValue;
+    }
+    return translatedValue.replace(this.templateMatcher,
+      (substring: string, b: string) => {
```

Comment:
What is "b"? Please use a more helpful parameter name.

Similarly, what is r below? This code is hard to understand.

Reply:
Changed!